src/yield/library.cairo

```diff
@@ -200,294 +114,54 @@ namespace CarbonableYielder {
-        }
-        return (new_remainder=remainder);
+        Claim.emit(address=caller, amount=amount, time=current_time);
+        return ();
     }
 }
 
 // Assert helpers
-namespace CarbonableYielder_assert {
-    func snapshoted{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
-        // [Check] Snapshot has been executed
-        let (time) = CarbonableYielder.snapshoted_time();
-        with_attr error_message("CarbonableYielder: provision must be executed after snapshot") {
-            assert_not_zero(time);
-        }
-        return ();
-    }
-
-    func snapshotable{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
-        // [Check] Previous provision has been executed
-        let (status) = CarbonableYielder_provisioned_.read();
-        with_attr error_message(
-                "CarbonableYielder: cannot snapshot if previous provision has not been set") {
-            assert status = TRUE;
-        }
-
-        // [Check] Snapshot timestamp is lower than the current time
-        let (current_time) = get_block_timestamp();
-        let (snapshoted_time) = CarbonableYielder.snapshoted_time();
-        with_attr error_message(
-                "CarbonableYielder: cannot snapshot at a sooner time that previous snapshot") {
-            assert_in_range(snapshoted_time, 0, current_time);
-        }
-
-        // [Check] Project is setup
-        let (carbonable_project_address) = CarbonableOffseter.carbonable_project_address();
-        let (slot) = CarbonableOffseter.carbonable_project_slot();
-        let (setup) = ICarbonableProject.isSetup(
-            contract_address=carbonable_project_address, slot=slot
-        );
-        with_attr error_message("CarbonableYielder: cannot snapshot if the project is not set up") {
-            assert setup = TRUE;
-        }
-
-        // [Check] At least 1 user registered
-        let (count) = CarbonableOffseter.total_user_count();
-        with_attr error_message(
-                "CarbonableYielder: cannot snapshot or provision if no user has registered") {
-            assert_not_zero(count);
-        }
-        return ();
-    }
-
-    func provisionable{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
-        // [Check] Provision has already been set
-        let (status) = CarbonableYielder_provisioned_.read();
-        with_attr error_message("CarbonableYielder: cannot provision if it has already been set") {
-            assert status = FALSE;
+namespace Assert {
+    func price_in_range{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        price: felt
+    ) {
+        let (contract_address) = CarbonableYielder_payment_token_address_.read();
+        let (decimals) = IERC20.decimals(contract_address=contract_address);
+        let unit = Math.pow(10, decimals);
+        with_attr error_message("CarbonableYielder: unexpected price") {
+            assert_in_range(price, 1 * unit, 1000 * unit + 1);
```

Comment:
Not sure but we might want to add prices equal to zero?

Reply:
Right, it was to ensure admin doesn't set 1 instead of 1000000 to mean 1$, but 0 should be accepted also. I'll remove the min threshold.